ducktape/cluster/vagrant.py

```diff
@@ -22,12 +26,49 @@ class VagrantCluster(JsonCluster):
     """
     An implementation of Cluster that uses a set of VMs created by Vagrant. Because we need hostnames that can be
     advertised, this assumes that the Vagrant VM's name is a routeable hostname on all the hosts.
+
+    - If cluster_file is specified in the constructor's kwargs (i.e. passed via command line argument --cluster-file)
+      - If cluster_file exists on the filesystem, read cluster info from the file
+      - Otherwise, retrieve cluster info via "vagrant ssh-config" from vagrant and write cluster info to cluster_file
+    - Otherwise, retrieve cluster info via "vagrant ssh-config" from vagrant
     """
 
-    def __init__(self):
-        hostname, ssh_hostname, username, flags = None, None, None, ""
-        nodes = []
+    def __init__(self, *args, **kwargs):
+        is_read_from_file = False
+
+        cluster_file = kwargs.get("cluster_file")
+        if cluster_file is not None:
+            try:
+                cluster_json = json.load(open(os.path.abspath(cluster_file)))
+                is_read_from_file = True
+            except IOError:
+                # It is OK if file is not found. Call vagrant ssh-info to read the cluster info.
+                pass
+
+        if not is_read_from_file:
+            cluster_json = {
+                "nodes": self._get_nodes_from_vagrant()
+            }
+
+        super(VagrantCluster, self).__init__(cluster_json)
+
+        is_aws = self._is_aws()
+        if not is_read_from_file:
+            # go through and find fully qualified domain name for each node
+            # this makes it possible to not require write access to /etc/hosts on the test driver machine
+            externally_routable_ip = {}
+            for node_account in self.available_nodes:
+                node_account.externally_routable_ip = self._externally_routable_ip(is_aws, node_account)
+                externally_routable_ip[str(node_account)] = node_account.externally_routable_ip
 
+            # If cluster_file is specified, cache cluster info in the file
+            if cluster_file is not None:
+                cluster_json["externally_routable_ip"] = externally_routable_ip
```

Comment:
Yep that's right we want this info but in slightly different format (hopefully my next comment will clarify this)